core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -45,27 +46,27 @@ private[log] case object LogCleaningPaused extends LogCleaningState
  *  While a partition is in the LogCleaningPaused state, it won't be scheduled for cleaning again, until cleaning is
  *  requested to be resumed.
  */
-private[log] class LogCleanerManager(val logDirs: Array[File], val logs: Pool[TopicPartition, Log]) extends Logging with KafkaMetricsGroup {
+private[log] class LogCleanerManager(val logDirs: ArrayBuffer[File], val logs: Pool[TopicPartition, Log]) extends Logging with KafkaMetricsGroup {
 
   import LogCleanerManager._
 
   override val loggerName = classOf[LogCleaner].getName
 
   // package-private for testing
   private[log] val offsetCheckpointFile = "cleaner-offset-checkpoint"
-  
+
   /* the offset checkpoints holding the last cleaned point for each log */
-  private val checkpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, offsetCheckpointFile)))).toMap
+  @volatile private var checkpoints = logDirs.map(dir => (dir, new OffsetCheckpointFile(new File(dir, offsetCheckpointFile)))).toMap
```

Comment:
I think it can be changed to a mutable map since a volatile var map can usually be replaced by a mutable map and vise versa. Does Kafka code has a preference between the two? I have chosen to use volatile var here because the code may be simpler. For example, I can simply do `checkpoints = checkpoints.filterKeys(_.getAbsolutePath != dir)` in `LogCleanerManager.handleLogDirFailure(dir: String)`.